File: Algo/Leetcode/152MaxProductSubarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int f_max = nums[0];
        int f_min = nums[0];
        int res = nums[0];
        for(int i=1; i<nums.size(); i++){
            int f_max_tmp = f_max; int f_min_tmp = f_min;
            f_max = max(max(f_max_tmp*nums[i], f_min_tmp*nums[i]), nums[i]);
            f_min = min(min(f_max_tmp*nums[i], f_min_tmp*nums[i]), nums[i]);
            res = max(f_max, res);
        }
        return res;
    }
};

File: Algo/Leetcode/152MaxProductSubarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        f_max = nums[0]
        f_min = nums[0]
        f = nums[0]
        for n in nums[1:]:
            f_max_tmp = f_max
            f_min_tmp = f_min
            f_max = max(n, f_max_tmp * n, f_min_tmp * n)
            f_min = min(n, f_max_tmp * n, f_min_tmp * n)
            f = max(f, f_max)

        return f